### ISSUE TITLE
yum_repository: remove unmanaged repo files

### ISSUE THREAD

**Issue by timss** [ansible/ansible-modules-extras#1680]
##### ISSUE TYPE

Feature Idea
##### COMPONENT NAME

yum_repository module
##### ANSIBLE VERSION

N/A
##### SUMMARY

The predecessor the the yumrepo module seems to be the [yumrepo role](https://github.com/picotrading/ansible-yumrepo), both made by @jtyr.

One feature I'm missing in the module is the ability to delete unmanaged repositories using the managed mode in the yumrepo role.

> The role also supports managed mode which removes any repo which is not managed by this role.

For anyone unfamiliar with the managed mode or the role itself, it creates `/etc/yum.repos.d/.managed` which includes the name of any repo you specified, and if enabled, deletes any repo not mentioned in this file at the end.

Using the module it is possible to remove existing repo files using `state: absent`. However, if you do not know which files are already present, or you plan to use the same playbook on hosts with different existing repo files it seems to me that this can quickly become a challenge to maintain.

Would it be possible to add a managed mode, or something similar to remove any repo files not specified using the yumrepo module? I might be able to help out, if that's desirable.

Thanks.

**Comment by jtyr** [comment]
The `yumrepo` role is actually creating a temp file for every YUM repo file it creates and then there is a final play which creates the `.managed` file composed of the names of the temp files. I could add the same feature into the `yumrepo` module but then it's up to the user to create the final play which will create the `.managed` file and deletes the unmanaged repos. I believe that's not possible to do on the module level. Just thinking, perhaps using Ansible facts to store the list of created repo files would be better than the temp files. Please feel free to elaborate on the possible solution.

**Comment by timss** [comment]
I have no previous experience developing Ansible modules, but looking at the yumrepo module code and official documentation it does indeed seem like a feasible solution to use facts (reference: [module provided facts](https://docs.ansible.com/ansible/developing_modules.html#module-provided-facts)). 

Not really in a position to say much about how it should be implemented at this point, but I'll try to take a look. My initial impression is that one could possibly add an option to `yumrepo` to trigger removal of repo files which are not available in the facts provided. If this sounds silly, please notify me. Thanks!

**Comment by jtyr** [comment]
> add an option to yumrepo to trigger removal of repo files which are not available in the facts provided

This is definitely possible but it must be called at the end of the playbook in order collect all the facts from all the previous `yumrepo` module calls.

**Comment by timss** [comment]
Yes, but I think that would be a reasonable design choice if the syntax is easy to use and comprehend.

**Comment by jtyr** [comment]
OK. Let me have a look how to implement this and once I have something I will link the PR to this issue in order you can test it.

**Comment by timss** [comment]
Great, thanks! 

I had a look myself, but was quickly stuck on how one should maintain a complete list of repos as a fact without overwriting it. Looking forward to see how you're going to solve it.

**Comment by jtyr** [comment]
I have looked at it and I'm afraid there is no solution to manage repos from the `yumrepos` module. The reason is that the module doesn't know about any of the variables/facts available in the Playbook or Role. The only input the module has are the module parameters. So in order to make the management working, we would have to pass the list of managed roles from the task to the module via parameter. The same applies for the `yumrepo_manage` variable. That means that every `yumrepo` module task would have to have those parameters defined. That's difficult to achieve if somebody is using a role from somebody else who did not parametrized it. The latest PR #1595 allows to add custom parameters (`params`) but still, not everybody provide this customization and even if they would, we would have to override it for every use of the `yumrepo` module which is quite uncomfortable.

**Comment by jtyr** [comment]
I just got another idea. We could update the `.managed` with the repo file name every time the module is called. Unmanaged files would be deleted only if the user would call the module with `state=managed`.

**Comment by jtyr** [comment]
@timss I think I nailed it. Please check the `EXAMPLES` section of the module documentation for an example how to use it. I would like to ask you to [test it](https://github.com/jtyr/ansible-modules-extras/blob/jtyr-yumrepo_managed/packaging/os/yumrepo.py) and provide your feedback before I create an official PR.

**Comment by timss** [comment]
Wow, great job! 

Code looks good, I'll have a go at it right away.

**Comment by jtyr** [comment]
There is one thing which I'm not sure about - if the `mgmt_enabled` should indicate that the `.managed` file was changed or if it should be always silent (return `changed=false`). What do you think?

I'm also thinking to to rename the states to `mm_enabled`, `mm_disabled` and the class to `ManagedMode`. Any opinion on this?

**Comment by timss** [comment]
Hm, having it return `changed=true` even if there is no change in the managed file or the repos can indeed be a bit misleading. If any metrics or monitoring uses the recap (ok/changed/unreachable/failed, ..) of an Ansible playbook which uses this module it will only add noise for repeated runs.

The user shouldn't really care about the `.managed` file, just whether some repos are managed or not. If any of the repos are changed, it'll report the change and I think that's enough. Having the 'enable()`function to return`changed=false` might be for the better.

:thumbsup: to `ManagedMode` / `mm` for me.

**Comment by jtyr** [comment]
> even if there is no change in the managed file

The thing is that the file has actually changed. The file is re-created every time the `mm_enabled` is called. This is why I wasn't sure if the module should report the change or not. But as you said, once we decide to use the managed mode, we do not care if the `.managed` file was changed. So I have changed it to only report the change in the `mm_ enabled` state if the file did not exist and was just created. Once it exists (and even if it gets modified) it won't report the change. Please test the changes.

**Comment by timss** [comment]
Sorry, my wording was a bit ambigious. I was thinking of change as in content of `.managed` file is the same before/after play, even though the file is re-created. Only reporting change if the file is new sounds like a good solution to me.

Tested your changes and it still works great. Output is good.

**Comment by jtyr** [comment]
Thanks for testing it. I will create an official PR now.

**Comment by timss** [comment]
Of course, glad to help! 

Once again thanks for looking into it. Looking at your code has been of great help in understanding Ansible modules.

You may want to add a note to your picotrading/ansible-yumrepo and/or the fork that it is now fully superseded by the yumrepo module. It's references in answers on Stack Overflow and a few other places.

For your information, I'm currently using it with Ansible 1.9.2.

**Comment by jtyr** [comment]
> You may want to add a note to your picotrading/ansible-yumrepo and/or the fork

Unfortunately I don't have write access to the picotrading repo any more but I will put a note into my [personal fork](https://github.com/jtyr/ansible-yumrepo) once it will be merged.

**Comment by timss** [comment]
Ah, ok. No biggie, your module will get much better exposure by being official and included in the documentation anyway :-)

I'll close this issue when the PR is merged.

**Comment by ansibot** [comment]
@jtyr, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@jtyr, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
This repository has been locked. All new issues and pull requests should be filed in https://github.com/ansible/ansible

Please read through the <a href="http://docs.ansible.com/ansible/dev_guide/repomerge.html#move-issues-and-prs-to-new-repo">repomerge</a> page in the dev guide. The guide contains links to tools which automatically move your issue or pull request to the ansible/ansible repo.

**Comment by ansibot** [comment]
This issue was migrated to https://github.com/ansible/ansible/issues/29273
